### ISSUE TITLE
Support enum constructs introduced in Python 3.11

### ISSUE THREAD

**Issue by hamdanal** [microsoft/pyright#6867]
Python 3.11 introduced, among other things in the enum module, [`enum.member` and `enum.nonmember`](https://docs.python.org/3/library/enum.html#enum.member) to explicitly declare enum class attributes as members or not. Pyright does not currently support these constructs:

```python
import enum
from typing import Literal, reveal_type

class E(enum.Enum):
    MEMBER = 1
    ANOTHER_MEMBER = enum.member(2)
    NON_MEMBER = enum.nonmember(3)

    @enum.member
    @staticmethod
    def ALSO_A_MEMBER() -> Literal[4]:
        return 4

def f(e: E) -> None:
    print(e)

f(E.MEMBER)
f(E.ANOTHER_MEMBER)
f(E.NON_MEMBER)  # no error: False negative
f(E.ALSO_A_MEMBER)  # error: False positive
try:
    print(f"{E.NON_MEMBER in E = }")  # no error: False negative
except TypeError as e:
    print(e)
print(f"{E.ALSO_A_MEMBER in E = }")

reveal_type(E.MEMBER.value)  # Type of "E.MEMBER.value" is "Literal[1]" - OK
reveal_type(E.ANOTHER_MEMBER.value)  # Type of "E.ANOTHER_MEMBER.value" is "member[int]" - Should be "Literal[2]"
try:
    reveal_type(E.NON_MEMBER.value)  # Type of "E.NON_MEMBER.value" is "nonmember[int]" - Should be an error
except AttributeError as e:
    print(e)
reveal_type(E.ALSO_A_MEMBER.value)  # Type of "E.ALSO_A_MEMBER.value" is "() -> Literal[4]" - OK
```

Running this file prints:
```
E.MEMBER
E.ANOTHER_MEMBER
3
E.ALSO_A_MEMBER
DeprecationWarning: in 3.12 __contains__ will no longer raise TypeError, but will return True or False depending on whether the value is a member or the value of a member
unsupported operand type(s) for 'in': 'int' and 'EnumType'
E.ALSO_A_MEMBER in E = True
Runtime type is 'int'
Runtime type is 'int'
'int' object has no attribute 'value'
Runtime type is 'staticmethod'
```

**Comment by erictraut** [comment]
AFAIK, no Python type checkers support these constructs yet. I wasn't even aware that they were added until now.

As an aside, the `enum` class in Python is full of weird and non-standard behaviors, and it seems that more of these oddities continue to be added every release. I really wish someone (perhaps the Python Steering Council?) would consider the broader impact of these decisions before allowing more of these oddities to be added. These seem to be added without any PEPs or opportunity for public comment. For example, did anyone think about how `enum.member` and `enum.nonmember` work with type stubs? Probably not.

**Comment by hamdanal** [comment]
> AFAIK, no Python type checkers support these constructs yet. I wasn't even aware that they were added until now.

I think part of the reason is that they were introduced recently, so unless you drop support for Python 3.10 in your code, you cannot use them. The other part is that they are niche features, you'll not need them in your everyday use of enum. TBH in the project where I am using `enum.nonmember`, we don't do any type checking in CI; I am more interested in Pylance support. Also, FWIW, there is a mypy issue open for new enum features in Python 3.11.

> As an aside, the `enum` class in Python is full of weird and non-standard behaviors, and it seems that more of these oddities continue to be added every release. I really wish someone (perhaps the Python Steering Council?) would consider the broader impact of these decisions before allowing more of these oddities to be added. These seem to be added without any PEPs or opportunity for public comment. For example, did anyone think about how `enum.member` and `enum.nonmember` work with type stubs? Probably not.

I didn't follow the details of how they were introduced and I kind of share your concerns here. There are a lot of stuff added to enum recently that I don't think I'll ever use. But for `enum.nonmember` specifically, I used it in three different places in my project already when I dropped support for Python 3.10 recently. This specific functionality is similar in principle to dataclass's `ClassVar` feature that allows you to define attribute that is not part of the dataclass fields.

**Comment by max-muoto** [comment]
Would agree with @hamanal. While it might be somewhat non-standard, I would definitely find use out of support here.

**Comment by erictraut** [comment]
This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.347, which I just published. It will also be included in a future release of pylance.
